### ISSUE TITLE
Wrong calculation heal when using Pugna's Ulti (heal amp. not working)

### ISSUE THREAD

**Issue by confeden** [ValveSoftware/Dota2-Gameplay#1832]
### Description

When Pugna's Ulti is cast on allies, the healing amount is not increased by Holy Locket, because Ulti cast on allies counts as dealing damage, not restoring health/mana.
It seems to me that this is the wrong spell mechanic. When used on an ally, the spell doesn't deal damage, but heals the ally, so it should count as a heal.

### Screenshots

<img width="584" alt="Screenshot 2022-07-16 115538" src="https://user-images.githubusercontent.com/7432044/179348100-6d9a8145-6c74-45ca-9fb5-cf7d4f11589c.png">

**Comment by confeden** [comment]
Allies' healing should not be reduced by magic resistance, and should also count as healing when receiving a bonus from Holy Locket (maybe I'm wrong, but I would like to see a response from the developers).

**Comment by Janexo27** [comment]
Seems like pipe on huskar affects pugna with 10% magic resistance. Healing done to teammates is equivalent to damage done to self. Increasing magic resistance on pugna decreases healing done to teammate. When you heal people you can also decrepify yourself you can heal more.

**Comment by confeden** [comment]
> Seems like pipe on huskar affects pugna with 10% magic resistance. Healing done to teammates is equivalent to damage done to self. Increasing magic resistance on pugna decreases healing done to teammate. When you heal people you can also decrepify yourself you can heal more.

When ally hero has Cloak for 550 gold, it also reduces heal amount from Pugna. It should not work in that way.

**Comment by TheUnholyWrath** [comment]
> > Seems like pipe on huskar affects pugna with 10% magic resistance. Healing done to teammates is equivalent to damage done to self. Increasing magic resistance on pugna decreases healing done to teammate. When you heal people you can also decrepify yourself you can heal more.
> 
> When ally hero has Cloak for 550 gold, it also reduces heal amount from Pugna. It should not work in that way.

Can you post screenshots of this? I can't reproduce it.

**Comment by confeden** [comment]
> Healing done to teammates is equivalent to damage done to self.

No, damage to itself is equal to heal amount (try to heal Antimage with 6 Pipe of Insight and neutral item with +20% resist = you will heal Antimage for 1-2 hp and damage Pugna with same 1-2 hp amount). 
In that case, heal amount decreased when ally trying to heal Antimage with 98% resist to 1-2 HP.
Awful mechanic with calculation heal amount.

**Comment by confeden** [comment]
> Can you post screenshots of this?

When I heal ally Centaur with item Pipe of Insight, he takes 49 heal amount, when he dropped Pipe of Insight, he receive 56 heal, it is because of magic resistance I think. 

<img width="1061" alt="Screenshot 2022-07-22 144928" src="https://user-images.githubusercontent.com/7432044/180433428-01da6c28-4afa-4b5b-b8c4-e2225bf85b70.png">

**Comment by TheUnholyWrath** [comment]
> > Can you post screenshots of this?
> 
> When I heal ally Centaur with item Pipe of Insight, he takes 59 heal amount, when he dropped Pipe of Insight, he receive 67 heal, it is because of magic resistance I think.
> 
> <img alt="Screenshot 2022-07-22 144928" width="1061" src="https://user-images.githubusercontent.com/7432044/180433428-01da6c28-4afa-4b5b-b8c4-e2225bf85b70.png">

Try any source of Magic Resistance except for Pipe

**Comment by confeden** [comment]
> Try any source of Magic Resistance except for Pipe

With neutral item Arcanist's Armor also reduce heal amount:
<img width="1135" alt="Screenshot 2022-07-22 145853" src="https://user-images.githubusercontent.com/7432044/180434745-abdf2fd8-182f-49bd-85b8-1e9e6809a751.png">

**Comment by TheUnholyWrath** [comment]
> > Try any source of Magic Resistance except for Pipe
> 
> With neutral item Arcanist's Armor also reduce heal amount: <img alt="Screenshot 2022-07-22 145853" width="1135" src="https://user-images.githubusercontent.com/7432044/180434745-abdf2fd8-182f-49bd-85b8-1e9e6809a751.png">

Okay cool, both Pipe and Arcanist's Armor have a Magic Resistance aura. Both of these auras are affecting the Pugna and increasing his Magic Resistance, so when he damages himself to heal his allies, he does less damage to himself. Thats why none of the other Magic Resistance items decrease it. If you give Pugna enough Cast Range to heal from outside the aura (aether + seer stone), you can see the values won't change.

**Comment by confeden** [comment]
Also very strange things: when I try to heal Antimage with 30 lvl + agh + shard (he has 73% magic resist) and Centaur with 30 lvl + shard (he has 25% magic resist) I get these numbers: (also Antimage and Centaur has empty inventory, without any items or aura's effect)
<img width="1247" alt="Screenshot 2022-07-22 150610" src="https://user-images.githubusercontent.com/7432044/180435793-c96de85a-81f5-4f60-bb5b-a2f55be4386f.png">

**Comment by TheUnholyWrath** [comment]
Okay you found another bug. Life Draining an ally with Aghs Scepter will increase Life Drains damage Values as if Pugna had a scepter.

**Comment by confeden** [comment]
> If you give Pugna enough Cast Range to heal from outside the aura (aether + seer stone), you can see the values won't change.

I got what you say:
<img width="1142" alt="Screenshot 2022-07-22 151422" src="https://user-images.githubusercontent.com/7432044/180437067-2a9459e0-a93d-4a41-ae79-2f7f5dd4cfd3.png">

But why I can't increase heal amount from item Holy Locket?
If Pugna has Holy Locket, he heals same value as without item, BUT when ally hero has item Holy Locket and Pugna trying to heal - it increase heals, but it should not work that, because Holy Locket should provide increased heal amount only for hero who wear it?

**Comment by confeden** [comment]
> When used on allies, Life Drain will increase Damage if the ally you are healing has a Scepter

Haha, nice I event don't know about that :DD

**Comment by confeden** [comment]
> When used on allies, Life Drain will increase Damage if the ally you are healing has a Scepter

same for item Holy Locket I think...

**Comment by TheUnholyWrath** [comment]
> > When used on allies, Life Drain will increase Damage if the ally you are healing has a Scepter
> 
> same for item Holy Locket I think...

I think so too, and its probably worth rewriting/retitling the issue now. Allies shouldn't need Scepter/Holy Locket for Life Drain damage/heal to increase. It seems like the exact same bug is happening for both items

**Comment by confeden** [comment]
It turns out very sadly: in order for Pugna to heal more of his allied hero, he must remove things from himself that increase magic damage :C
I don't think the game mechanics should be like that. Maybe it should be changed for Pugna?

**Comment by confeden** [comment]
I think that the following mechanic will be correct: heal an allied hero for a value of (for example) +60, regardless of what magic resistance Pugna has, but Pugna should take damage depending on magic resistance.
At the moment, the amount of the allied hero's healing depends on Pugna's magic resistance, and this is not correct.

**Comment by confeden** [comment]
When Pugna has 93% magic resistance, he definitely lose options for healing ally Hero:
<img width="982" alt="Screenshot 2022-07-22 155430" src="https://user-images.githubusercontent.com/7432044/180443590-327f9804-9b68-4e5d-93b8-b568a34bc792.png">

**Comment by confeden** [comment]
Also, when ally Oracle cast spell Fate's Edict to ally Pugna, it grants Pugna 100% magic resist so 5 seconds Pugna just sucking and no one are happy :D
with 100% magic resistance Pugna can't heal ally, is should work other way, please developers look at this :C
<img width="1106" alt="Screenshot 2022-07-22 155959" src="https://user-images.githubusercontent.com/7432044/180444856-39cb19ad-3522-443b-9a92-d92511f90e46.png">
